fuzz_lightyear/generator.py

```diff
@@ -96,26 +105,63 @@ def _add_request_to_sequence(
     for sequence in seed:
         for tag_group in dir(client):
             for request in _generate_requests(tag_group):
-                # Attempt to add every single permutation first.
-                new_sequence = sequence + [request]
-
-                # TODO: sequence pruning. We would get a lot better results if
-                #       we only added a request that consumes resources that
-                #       have been created by the sequence so far. However, due
-                #       to the nature of microservices, we may need to rely on
-                #       factories for this.
-                #
-                #       This is further made complicated because we have no idea
-                #       whether a request should consume an existing resource,
-                #       use a factory, or fuzz a value.
-                #
-                #       This means, we need to do all permutations, because we
-                #       can't possibly know whether to eagerly prune.
-                output.append(new_sequence)
+                neighbor_requests = request_graph[request.operation_id]
+                # We have no idea whether a request should consume an existing
+                # resource, use a factory, or fuzz a value.
+                # However, we want to move towards building sequences of requests
+                # that only consume resources created by previous requests (the
+                # algorithm described in the original Microsoft paper)
+                # Therefore, add the request to the sequence if it shares an edge with
+                # ANY of the requests in the sequence- leaving room for factory
+                # usage, but still keeping in line with the spirit of RESTler.
+                if any([req.operation_id in neighbor_requests for req in sequence]):
+                    new_sequence = sequence + [request]
+                    output.append(new_sequence)
 
     return output
 
 
+@lru_cache(maxsize=1)
+def _generate_request_graph() -> Dict[str, set]:
+    """
+    Generates a directed graph, represented as an adjacency list.
+
+    Every node in the graph is an operation_id, and an edge between two nodes
+    A and B means that node A consumes a resource that node B produces.
+
+    """
+    client = get_abstraction().client
+    produces = defaultdict(list)  # type: Dict[str, List[str]]
+    consumes = defaultdict(list)  # type: Dict[str, List[str]]
```

Comment:
For `consumes`, or for both? I think we'd want to keep `produces` as a `defaultdict` because we are appending to the list of operations that correspond to a param (line 146), but I agree that `consumes` does not need to be a `defaultdict`.

Reply:
Yeah, I was referencing the `consumes` case.